mini-projects/01-device-driver/laptop-interface/src/gui.rs

```diff
@@ -0,0 +1,390 @@
+use std::time::Duration;
+
+use crate::controller_management::{BackgroundMultiDevice, ControllerInput, DeviceState};
+use eframe::egui;
+use egui::{Frame, Margin};
+use plotters::{
+    backend::BitMapBackend,
+    chart::ChartBuilder,
+    prelude::{Circle, IntoDrawingArea},
+    series::LineSeries,
+    style::{BLUE, RED, WHITE},
+};
+
+use std::path::Path;
+use std::time::Instant;
+
+// Struct representing the main application, managing controller states and game logic
+pub struct GameApp {
+    device_manager: Option<BackgroundMultiDevice>,
+    winner: Option<String>,
+    time: Instant,
+    controller_one_pos: Vec<(f64, (i32, i32))>,
+    controller_two_pos: Vec<(f64, (i32, i32))>,
+}
+
+// Default configuration for GameApp, initializing fields
+impl Default for GameApp {
+    fn default() -> Self {
+        Self {
+            device_manager: None,
+            winner: None,
+            time: Instant::now(),
+            controller_one_pos: vec![(0.0, (0, 0))],
+            controller_two_pos: vec![(0.0, (0, 0))],
+        }
+    }
+}
+
+#[allow(dead_code)]
+impl GameApp {
+    pub fn get_state(&self) -> DeviceState {
+        self.device_manager.as_ref().unwrap().get_state()
+    }
+    pub fn get_pos_player_one(&self) -> (i32, i32) {
+        self.device_manager.as_ref().unwrap().get_pos()[0]
+    }
+    pub fn get_pos_player_two(&self) -> (i32, i32) {
+        self.device_manager.as_ref().unwrap().get_pos()[1]
+    }
+    pub fn set_controller_input(&mut self, controller_input: Option<ControllerInput>) {
+        self.device_manager
+            .as_mut()
+            .unwrap()
+            .set_controller_input(controller_input);
+    }
+
+    fn plot_data(
```

Comment:
nit: in general, this function probably could have been more generic such that it could handle x controllers - I know y'all didn't tackle that ticket, but still good to keep in mind when architecting interfaces